### ISSUE TITLE
Jenkins runner support

### ISSUE THREAD

**Issue by migmartri** [chainloop-dev/chainloop#88]
Chainloop [has support for CI/CD runners](https://docs.chainloop.dev/reference/operator/contract#runner-context), this task is about adding Jenkins (and JenkinsX) to the list of supported ones.

On the technical side it meens implementing this interface https://github.com/chainloop-dev/chainloop/blob/053637d91d5aa728d50ec3740f246fdca8120b87/internal/attestation/crafter/runner.go#L27-L37

which will require you to inspect and figure out 1) how to know that you are in that runner, 2) what env variables you care about for the tool to auto-discover and 3) how to create a `link` to the workflow.

Additionally you'll need to update the `contract` definition to support this new runner type 

https://github.com/chainloop-dev/chainloop/blob/053637d91d5aa728d50ec3740f246fdca8120b87/app/controlplane/api/workflowcontract/v1/crafting_schema.proto#L36-L40

**Comment by buccarel** [comment]
Looking after this.

**Comment by buccarel** [comment]
This is interesting:

During the installation of jenkins (in this particular case, I'm using a docker container), the following setup step happens through the guided procedure:

<img width="971" alt="image" src="https://github.com/chainloop-dev/chainloop/assets/11213785/eb87581f-e31c-4069-aa0c-8e6f774ec059">

I will need to double-check it but as the wizard says this configuration is *required*, most likely the RunURI() method can be implemented using these values. I am assuming this could be retrieved from the job's node.

**Comment by migmartri** [comment]
Correct, I believe that URL matches `JENKINS_URL` env var

https://e.printstacktrace.blog/jenkins-pipeline-environment-variables-the-definitive-guide/

**Comment by buccarel** [comment]
To make sure, I am creating an alias to 127.0.0.1 that is different than localhost. I will use such alias to configure jenkins, to see if it appears in the job's node. That will be the ultimate confirmation.

**Comment by buccarel** [comment]
Just some notes, mostly for myself.

If we SSH into a jenkins agent node, we can see that the info available at system level is not much:

<img width="1207" alt="image" src="https://github.com/chainloop-dev/chainloop/assets/11213785/53a0b557-f38d-4b59-953b-93cb85fed43d">

The env variable `AGENT_WORKDIR` should be guaranteed to be present.

If I launch a job on the node, the set of running processes and the set of environment variables do not change at all. This means that the Groovy process is ran inside `agent.jar`, which by the way should also be guaranteed to be present on the node, and should have that particular name. Specifically, its pathname should end with the suffix `jenkins/agent.jar`.

Instead, if we are the jenkins job, we have a lot more info at our disposal. The set of environment variables is the following:
```
JENKINS_HOME=/var/jenkins_home
CI=true
RUN_CHANGES_DISPLAY_URL=http://jenkins-master:8080/job/simple-pipeline/9/display/redirect?page=changes
HOSTNAME=a5bcfe8f5f10
NODE_LABELS=root-inbound-agent docker
HUDSON_URL=http://jenkins-master:8080/
HOME=/home/jenkins
BUILD_URL=http://jenkins-master:8080/job/simple-pipeline/9/
HUDSON_COOKIE=f0961d75-4083-4ace-9ea6-4f7ea5a48716
JENKINS_SERVER_COOKIE=durable-eda971e81cf14777d9b1fed11303b0dda56bb649f09a6e05da324f4698fa9cc5
WORKSPACE=/home/jenkins/agent/workspace/simple-pipeline
NODE_NAME=root-inbound-agent
RUN_ARTIFACTS_DISPLAY_URL=http://jenkins-master:8080/job/simple-pipeline/9/display/redirect?page=artifacts
STAGE_NAME=Inspect environment
EXECUTOR_NUMBER=0
user=jenkins
RUN_TESTS_DISPLAY_URL=http://jenkins-master:8080/job/simple-pipeline/9/display/redirect?page=tests
BUILD_DISPLAY_NAME=#9
HUDSON_HOME=/var/jenkins_home
AGENT_WORKDIR=/home/jenkins/agent
JOB_BASE_NAME=simple-pipeline
PATH=/opt/java/openjdk/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
BUILD_ID=9
BUILD_TAG=jenkins-simple-pipeline-9
LANG=C.UTF-8
JENKINS_URL=http://jenkins-master:8080/
JOB_URL=http://jenkins-master:8080/job/simple-pipeline/
BUILD_NUMBER=9
JENKINS_NODE_COOKIE=b6209db8-1376-4931-89e7-2b9a46862332
RUN_DISPLAY_URL=http://jenkins-master:8080/job/simple-pipeline/9/display/redirect
HUDSON_SERVER_COOKIE=a4d8db8ebe73e068
JOB_DISPLAY_URL=http://jenkins-master:8080/job/simple-pipeline/display/redirect
JOB_NAME=simple-pipeline
JAVA_HOME=/opt/java/openjdk
PWD=/home/jenkins/agent/workspace/simple-pipeline
WORKSPACE_TMP=/home/jenkins/agent/workspace/simple-pipeline@tmp
TZ=Etc/UTC
```
The set of running processes is much richer than before. I am not pasting it here because the environemtn variables are going to be more than enough to guarantee that we can detect a jenkins node.

**Comment by buccarel** [comment]
In my opinion, to detect that we are in a jenkins job, we could check the presence of all of the following environment variables:
* JENKINS_HOME
* BUILD_URL
* NODE_NAME
* AGENT_WORKDIR

As optional, we could also detect that the process `agent.jar` is running. I don't think it's necessary.

**Comment by migmartri** [comment]
Thanks for digging into this and sharing your findings. 

Your suggestion makes sense to me, although let me share some thoughts in case you think it makes sense.

## Detection via `CheckEnv()`

Note that this feature is not a security enforcement but more like a duck-typing kind of detection. So, to me the minimum we could add, the better.

So far, for other runner types, my choice was based on 

- a) what's something that could unequivocally identify this runner type and
- b) what other env vars do we need to generate the JOB URL

So it seems that in your case, we'd have enough with making sure that both `JENKINS_HOME` + `BUILD_URL` are present, don't you think?

## Env Vars listing `ListEnvVars()`

This method is used to automatically inject env vars into the attestation. I'd recommend to check what variables we use in other runners to support a similar context. 

Long story short, what we try to add is some representation of:

- Who triggered the job (in your case maybe the `user` env var)
- Some information about the commit (I can't see that in your vars, it might not exist since it was not triggered automatically by a change?)
- Some information about the agent where the job is running (you have `AGENT_WORKDIR` and maybe `NODE_NAME`)

This is for example what we do for GitHub actions.

https://github.com/chainloop-dev/chainloop/blob/053637d91d5aa728d50ec3740f246fdca8120b87/internal/attestation/crafter/runners/githubaction.go#L42-L52

In any case, feel free to create a PR with your suggestions, there is not right or wrong answer here and we can iterate later on! 

THank you!!

**Comment by buccarel** [comment]
Yep, I think you're very right. I was going to use those 2 vars only :P

I didn't check Jenkins X yet, though. Is it a problem if I discover that Jenkins X instantiate the same 2 variables?

**Comment by migmartri** [comment]
> I didn't check Jenkins X yet, though. Is it a problem if I discover that Jenkins X instantiates the same two variables?

I'm not 100% sure, but I think Jenkins and JenkinsX are different enough to be implemented as different runner types. So, ideally, a Jenkins X run should not impersonate a regular Jenkins run. 

That said, and after looking at [this JenkinsX reference doc](https://jenkins-x.io/v3/develop/reference/variables/), your proposal seems to work as expected.

**Comment by buccarel** [comment]
From what I read yesterday about Jenkins X, I share the feeling that they should have different runners.
I still don't know very well what Jenkins X is, so I was planning to do cover it afterwards in another PR, unless you wish otherwise.

**Comment by migmartri** [comment]
> From what I read yesterday about Jenkins X, I share the feeling that they should have different runners. I still don't know very well what Jenkins X is, so I was planning to do cover it afterwards in another PR, unless you wish otherwise.

Makes sense to me!

Thanks!
